File: 3243_shortest_distance_after_road_addition_queries_i.cpp
#include <vector>
#include <queue>
#include <numeric>
using namespace std;

class Solution {
public:
    vector<int> shortestDistanceAfterQueries(int g, vector<vector<int>>& h) {
        vector<int> i, j(g);
        vector<vector<int>> k(g);
        iota(j.begin(), j.end(), 0);
        for (int l = 0; l < g - 1; l++) k[l].push_back(l + 1);
        for (auto m : h) {
            int n = m[0], o = m[1];
            k[n].push_back(o);
            if (j[n] + 1 < j[o]) {
                queue<int> p;
                p.push(o);
                j[o] = j[n] + 1;
                while (p.size()) {
                    int q = p.front();
                    p.pop();
                    for (auto r : k[q]) {
                        if (j[q] + 1 < j[r]) {
                            j[r] = j[q] + 1;
                            p.push(r);
                        }
                    }
                }
            }
            i.push_back(j.back());
        }
        return i;
    }
};

File: 3243_shortest_distance_after_road_addition_queries_i.java
import java.util.*;

class Solution {
    public int[] shortestDistanceAfterQueries(int a, int[][] b) {
        List<Integer> c = new ArrayList<>();
        int[] d = new int[a];
        List<List<Integer>> e = new ArrayList<>();
        for (int f = 0; f < a; f++) {
            d[f] = f;
            e.add(new ArrayList<>());
        }
        for (int f = 0; f < a - 1; f++) {
            e.get(f).add(f + 1);
        }
        for (int[] g : b) {
            int h = g[0], i = g[1];
            e.get(h).add(i);
            if (d[h] + 1 < d[i]) {
                Queue<Integer> j = new LinkedList<>();
                j.add(i);
                d[i] = d[h] + 1;
                while (!j.isEmpty()) {
                    int k = j.poll();
                    for (int l : e.get(k)) {
                        if (d[k] + 1 < d[l]) {
                            d[l] = d[k] + 1;
                            j.add(l);
                        }
                    }
                }
            }
            c.add(d[a - 1]);
        }
        return c.stream().mapToInt(x -> x).toArray();
    }
}
